utils/zerovec/src/zerovec/mod.rs

```diff
@@ -257,7 +257,9 @@ where
     /// `bytes` need to be an output from [`ZeroSlice::as_bytes()`].
     pub const unsafe fn from_bytes_unchecked(bytes: &'a [u8]) -> Self {
         // &[u8] and &[T::ULE] are the same slice with different length metadata.
-        let (data, mut metadata): (usize, usize) = core::mem::transmute(bytes);
```

Comment:
FWIW the old code most likely has UB. Quoting from the `transmute` docs:

> Transmuting pointers to integers in a const context is [undefined behavior](https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html). Any attempt to use the resulting value for integer operations will abort const-evaluation.

So, if `from_bytes_unchecked` is ever called with a non-zero-sized slice (or a zero-sized slice that actually uses a pointer to some allocated memory), then this is definitely UB. This PR is a UB-fix, not a rustc-regression-workaround.

Reply:
Yeah I'm aware it's UB.

As I've stated in https://github.com/rust-lang/rust/issues/99923, I still consider things like this a regression as long as the docs on undefined behavior are incomplete, scattered, and in flux.